core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala

```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
+      // TODO: Unflattened INFO field splitting
+      logger.warn(
+        "The variant splitting transformer does not split unflattened INFO fields within the attributes column.")
+    }
+
+    // Update splitFromMultiAllelic column, add INFO_OLD_MULTIALLELIC column (see vt decompose)
+    // and posexplode alternateAlleles column
+    val dfAfterAltAlleleSplit = variantDf
+      .withColumn(
+        splitFromMultiAllelicField.name,
+        when(size(col(alternateAllelesField.name)) > 1, lit(true)).otherwise(lit(false))
+      )
+      .withColumn(
+        infoFieldPrefix + oldMultiallelicFieldName,
+        when(
+          col(splitFromMultiAllelicField.name),
+          concat(
+            col(contigNameField.name),
+            lit(":"),
+            expr(s"${startField.name} + 1"),
+            lit(":"),
+            concat_ws("/", col(s"${refAlleleField.name}"), col(s"${alternateAllelesField.name}"))
+          )
+        ).otherwise(lit(null))
+      )
+      .select(
+        col("*"),
+        posexplode(col(alternateAllelesField.name))
+          .as(Array(splitAlleleIdxFieldName, splitAllelesFieldName))
+      )
+
+    // Split INFO fields
+    val dfAfterInfoSplit = splitInfoFields(dfAfterAltAlleleSplit)
+
+    // split genotypes fields, update alternateAlleles field, and drop the columns resulting from posexplode
+    splitGenotypeFields(dfAfterInfoSplit)
+      .withColumn(alternateAllelesField.name, array(col(splitAllelesFieldName)))
+      .drop(splitAlleleIdxFieldName, splitAllelesFieldName)
+  }
+
+  /**
+   * Generates a new DataFrame by splitting the info fields, based on splitAlleleIdx
+   * field generated by posexplode
+   * @param variantDf
+   * @return dataframe with split info fields
+   */
+  @VisibleForTesting
+  private[normalizevariants] def splitInfoFields(variantDf: DataFrame): DataFrame = {
+    variantDf
+      .schema
+      .filter(field =>
+        field.name.startsWith(infoFieldPrefix) && field.dataType.isInstanceOf[ArrayType])
+      .foldLeft(
+        variantDf
+      )(
+        (df, field) =>
+          df.withColumn(
+            field.name,
+            when(
+              col(splitFromMultiAllelicField.name) &&
+              size(col(field.name)) === size(col(alternateAllelesField.name)),
+              array(expr(s"${field.name}[$splitAlleleIdxFieldName]"))
+            ).otherwise(col(field.name))
+          )
+      )
+  }
+
+  /**
+   * Generates a new DataFrame by splitting the genotypes subfields, based on splitAlleleIdx
+   * field generated by posexplode
+   * @param variantDf
+   * @return dataframe with split genotype subfields
+   */
+  @VisibleForTesting
+  private[normalizevariants] def splitGenotypeFields(variantDf: DataFrame): DataFrame = {
+
+    // get genotypes schema
+    val genotypesFieldDataType = variantDf
+      .schema
+      .fields
+      .find(_.name == genotypesFieldName)
+      .get
+      .dataType
+
+    if (!genotypesFieldDataType.isInstanceOf[ArrayType]) {
```

Comment:
I think that @karenfeng added a utility function to validate and get the genotype schema.

Reply:
You can find this in `InternalRowToVariantContextConverter.getGenotypeSchema`

Reply:
Replaced.